src/auth/envDetect.ts

```diff
@@ -0,0 +1,65 @@
+/**
+ * Copyright 2018 Google LLC. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as gcpMetadata from 'gcp-metadata';
+
+export enum GCPEnv {
+  APP_ENGINE,
+  KUBERNETES_ENGINE,
+  CLOUD_FUNCTIONS,
+  COMPUTE_ENGINE,
+  NONE
+}
+
+let env: GCPEnv;
+
+export async function getEnv() {
+  if (!env) {
```

Comment:
Note that for the `APP_ENGINE` case, `env` will be falsy, meaning that the cache doesn't work on App Engine.

Reply:
```ts
env = GCPEnv.APP_ENGINE;
if (env) {
  console.log('env is truthy');
} else {
  console.log('env is falsy'); // <<-- this is printed
}
```

The reason is that `APP_ENGINE` will correspond to the value 0 being the first thing in the enum. 0 is falsy.